napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,643 @@
+                    self.color_properties.values, selected_indices
+                )
+                self.color_properties = ColorProperties(
+                    name=color_property_name,
+                    values=new_color_property_values,
+                    current_value=current_value,
+                )
+
+    def paste(self, colors: np.ndarray, properties: Dict[str, np.ndarray]):
+        """Append colors to the ColorManager. Uses the color values if
+        in direct mode and the properties in colormap or cycle mode.
+
+        This method is for compatibility with the paste functionality
+        in the layers.
+
+        Parameters
+        ----------
+        colors : np.ndarray
+            The (Nx4) color array of color values to add. These values are
+            only used if the color mode is direct.
+        properties : Dict[str, np.ndarray]
+            The property values to add. These are used if the color mode
+            is colormap or cycle.
+        """
+        if self.mode == ColorMode.DIRECT:
+            self.colors = np.concatenate(
+                (self.colors, transform_color(colors))
+            )
+        else:
+            color_property_name = self.color_properties.name
+            current_value = self.color_properties.current_value
+            old_properties = self.color_properties.values
+            values_to_add = properties[color_property_name]
+            new_color_property_values = np.concatenate(
+                (old_properties, values_to_add),
+                axis=0,
+            )
+
+            self.color_properties = ColorProperties(
+                name=color_property_name,
+                values=new_color_property_values,
+                current_value=current_value,
+            )
+
+    def update_current_properties(
+        self, current_properties: Dict[str, np.ndarray]
+    ):
+        """This is updates the current_value of the color_properties when the
+        layer current_properties is updated.
+
+        This is a convenience method that is generally only called by the layer.
+
+        Parameters
+        ----------
+        current_properties : Dict[str, np.ndarray]
+            The new current property values
+        """
+        if self.color_properties is not None:
+            current_property_name = self.color_properties.name
+            current_property_values = self.color_properties.values
+            if current_property_name in current_properties:
+                new_current_value = np.squeeze(
+                    current_properties[current_property_name]
+                )
+                if new_current_value != self.color_properties.current_value:
+                    self.color_properties = ColorProperties(
+                        name=current_property_name,
+                        values=current_property_values,
+                        current_value=new_current_value,
+                    )
+                    self.events.colors()
```

Comment:
I agree that we should take advantage of the automated event emission. I believe the issue here is that since we are setting `ColorManager.color_properties` and then the `root_validator`handles the `ColorManager.colors` update, the `colors` event isn't' emitted (i.e., the `ColorManager.colors` setter isn't used).

Upon further though, I believe that linking the `color_properties` event to the vispy update would allow us to avoid manually emitting events.

Reply:
I connected the `color_properties` event in the vispy file, so we no longer need the manual event emission.